### DISCUSSION TITLE
Z Offset adjustment while printing.

### DISCUSSION THREAD

**Discussion by jgbrown54** [Guizz27/DWIN_LK4Pro#20] [General]
One user is complaining that when he goes into the Z Offset screen while printing the first layer, that reguardless of the current Z Offset setting the display shows a 0.0 and he is unable to change it. Any suggestions?

**Comment by Guizz27** [comment]
This is likely that he does not have a BLTouch. The Z Offset adjustment while printing only works with a BLTouch because it sets the Z-offset which only exists with a mesh bad leveling to compensate for the probe not triggering at height 0.

I have 2 options: either I hide the z-offset controls when no BLTouch or I make it adjust the baby steps and display the proper value on the screen.

**Comment by jgbrown54** [comment]
That makes perfect sense. I just didn't think of it. What is a Z Probe Offset if you don't have a Z Probe. It would be nice though if our non-BLTouch users could adust the probe height during the first layer. But then where would you save it and when would it reset? Will Marlin allow saving it and using it if you don't have a Mesh? 

Can we use Mesh Bed Leveling?

MESH_BED_LEVELING
 *   Probe a grid manually
 *   The result is a mesh, suitable for large or uneven beds. (See BILINEAR.)
 *   For machines without a probe, Mesh Bed Leveling provides a method to perform
 *   leveling in steps so you can manually adjust the Z height at each grid-point.
 *   With an LCD controller the process is guided step-by-step.

**Comment by Guizz27** [comment]
The control on the screen is meant as babystepping
https://marlinfw.org/docs/gcode/M290.html

Which with a probe, it can also modify the Z-offset. Otherwise, I don`t think baby step value is meant to be saved between prints. 

It is labelled as "Z-offset" on the screen which is actually correct with a probe. The problem is without a probe. Either I remove the control or I label it "Baby Step + Z-offset" and fix the code to work without a probe.

Manual mesh leveling could be added but would not fix this Z-offset control in the adjust screen since Z-offset only makes sense for a probe. I misused the word "mesh bed leveling" in my previous comment. I meant automatic bed leveling which also creates a mesh but automatically with a probe.

**Comment by Guizz27** [comment]
Well I just looked at the code.
When modifying the value on the screen, baby steps are adjusted and if you have a probe, the z-offset is also adjusted and saved.
The problem is the screen displays the z-offset which is 0 if you don't have a probe.

baby step does not have a value. It is always added to the current z position. If you enter 1mm, it will add 1mm to the z axis and move up by 1mm. If you set 1mm again, it goes up another 1mm. It is basically a "Move while printing".

With a probe, the z-offset is subtracted from the value you entered, and the difference is added to the z and the z-offset. Which means if you enter 1mm and your z-offset is -1mm, it will move up by 2mm (from -1 to 1) and set the z-offset to (-1 + 2 = 1). If you enter again 1mm it won't move because 1mm - 1mm = 0;

**Comment by jgbrown54** [comment]
What if....For ABL you leave it as Z Offset. But for Non-ABL you made it Z Height? You could display the curent height and allow them to raise or lower it.

What if the ABL and Non-ABL is on layer 3 and they change the value? Can they or only on layer 1? Or what if they lower it below the current layer height? Or what if they enter the manual input screen and before they enter a value, the layer changes?

**Comment by Guizz27** [comment]
BaBy Step moves the Z but does not change the value. It doesn't care where the Z or layer (it doesn't know about layers). So if you are at Z 0.6, you enter 1mm, the Z will move up 1 mm but the Z value is still 0.6. The values you enter there would normally be hundredths or tenths of a mm.

What I will do is start the Baby Step at 0 when you start a print. When you modify it, I keep a copy, display it on the screen. Then when you enter a new value, I can do a diff with the previous and move the Z by the delta. This would be sort of a Z-offset but initialized at 0 when the print starts for non-BLTouch.

**Comment by jgbrown54** [comment]
That sounds like it would work if possible. And it would add functionality for people that don't have ABL. But we are talking about the Z Probe Offset when they don't have a probe. So whatever you decide.
